File: src/main/java/net/hectus/neobb/turn/default_game/structure/TIronBarJail.java
package net.hectus.neobb.turn.default_game.structure;

import net.hectus.neobb.buff.Buff;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.structure.PlacedStructure;
import net.hectus.neobb.structure.Structure;
import net.hectus.neobb.structure.StructureManager;
import net.hectus.neobb.turn.default_game.CounterFilter;
import net.hectus.neobb.turn.default_game.attributes.clazz.NeutralClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.RedstoneClazz;
import net.hectus.neobb.turn.default_game.attributes.clazz.SupernaturalClazz;
import net.hectus.neobb.turn.default_game.attributes.function.BuffFunction;
import net.hectus.neobb.turn.default_game.attributes.function.CounterFunction;
import net.hectus.neobb.util.Modifiers;

import java.util.List;

public class TIronBarJail extends StructureTurn implements BuffFunction, NeutralClazz {
    protected NeoPlayer trapped = null;

    public TIronBarJail(NeoPlayer player) { super(player); }
    public TIronBarJail(PlacedStructure data, NeoPlayer player) {
        super(data, player);
        this.trapped = player.nextPlayer();
    }

    @Override
    public int cost() {
        return 7;
    }

    @Override
    public Structure referenceStructure() {
        return StructureManager.structure("iron_bar_jail");
    }

    @Override
    public void apply() {
        // This is not actually checking if the player is INSIDE it, but this is easier and works good enough.
        if (trapped == null || trapped.location().distance(data.lastBlock().getLocation()) > 1.5) return;

        trapped.addModifier(Modifiers.P_NO_MOVE);
        trapped.addModifier(Modifiers.P_DEFAULT_NO_ATTACK);

        player.game.turnScheduler.runTaskTimer("iron_bar_jail", () -> {
            if (player.game.history().getLast() instanceof CounterFunction &&
                    CounterFilter.clazz(RedstoneClazz.class).doCounter(player.game.history().getLast()) ||
                    CounterFilter.clazz(SupernaturalClazz.class).doCounter(player.game.history().getLast())) {
                trapped.removeModifier(Modifiers.P_NO_MOVE);
                trapped.removeModifier(Modifiers.P_DEFAULT_NO_ATTACK);
            }
        }, () -> !trapped.hasModifier(Modifiers.P_NO_MOVE) || !trapped.hasModifier(Modifiers.P_DEFAULT_NO_ATTACK), 1);
    }

    @Override
    public List<Buff> buffs() {
        return List.of();
    }
}

File: src/main/kotlin/net/hectus/neobb/modes/turn/default_game/structure/TIronBarJail.kt
package net.hectus.neobb.modes.turn.default_game.structure

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.buff.Buff
import net.hectus.neobb.game.util.ScheduleID
import net.hectus.neobb.matrix.structure.PlacedStructure
import net.hectus.neobb.matrix.structure.StaticStructure
import net.hectus.neobb.matrix.structure.StaticStructures
import net.hectus.neobb.modes.turn.default_game.CounterFilter
import net.hectus.neobb.modes.turn.default_game.attribute.clazz.NeutralClazz
import net.hectus.neobb.modes.turn.default_game.attribute.clazz.RedstoneClazz
import net.hectus.neobb.modes.turn.default_game.attribute.clazz.SupernaturalClazz
import net.hectus.neobb.modes.turn.default_game.attribute.function.BuffFunction
import net.hectus.neobb.modes.turn.default_game.attribute.function.CounterFunction
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.util.Modifiers

class TIronBarJail(data: PlacedStructure?, cord: Cord?, player: NeoPlayer?) : StructureTurn(data, cord, player), BuffFunction, NeutralClazz {
    override val cost: Int = 7
    override val staticStructure: StaticStructure = StaticStructures.Default.IRON_BAR_JAIL

    private val trapped: NeoPlayer?
        get() = player?.nextPlayer()

    override fun apply() {
        // This is not checking if the player is INSIDE it, but this is easier and works well enough.
        if (trapped == null || trapped!!.location().distance(data!!.lastBlock.location) > 1.5) return

        trapped!!.addModifier(Modifiers.Player.NO_MOVE)
        trapped!!.addModifier(Modifiers.Player.Default.NO_ATTACK)

        player!!.game.turnScheduler.runTaskTimer(ScheduleID.IRON_BAR_JAIL, 1, {
            !trapped!!.hasModifier(Modifiers.Player.NO_MOVE) || !trapped!!.hasModifier(Modifiers.Player.Default.NO_ATTACK)
        }) {
            if (player.game.history.last() is CounterFunction &&
                CounterFilter.clazz(RedstoneClazz::class).doCounter(player.game.history.last()) ||
                CounterFilter.clazz(SupernaturalClazz::class).doCounter(player.game.history.last())
            ) {
                trapped!!.removeModifier(Modifiers.Player.NO_MOVE)
                trapped!!.removeModifier(Modifiers.Player.Default.NO_ATTACK)
            }
        }
    }

    override fun buffs(): List<Buff<*>> = emptyList()
}
